Run end-to-end tests in Espresso mode

jbearer opened the pull request:
Depends on https://github.com/EspressoSystems/espresso-sequencer/pull/570
Depends on https://github.com/EspressoSystems/op-geth/pull/2

Closes #31 
Closes #30 
Closes #43 
Closes #8 
Closes #7

jbearer committed:
Set up e2e tests to use Espresso mode

Running the tests with both legacy mode and Espresso mode in the
same process is problematic, because the e2e tests use global
variables initialized based on command line arguments to get the
system config and contract deployments, both of which must change
depending on the mode.

Instead, the test code does not change at all, and we invoke
`go test` twice, with different command line arguments, to run the
tests using an Espresso deployment and a legacy deployment.

The devnet Python script has been parameterized based on a directory
for output files, and there is a new make target to generate
contract deployments and L1 genesis state with the `Espresso` flag
enabled.

CircleCI jobs have been added to run the e2e tests in Espresso mode.
These jobs will not pass until we finish implementing the sequencer
(https://github.com/EspressoSystems/op-espresso-integration/issues/30)
and then we will probably have to selectively enable/disable certain
tests that we want to run with Espresso.

jbearer committed:
Clean up sequencer logging

jbearer committed:
Declare CircleCI parameters

jbearer committed:
Fix CircleCI dependencies

jbearer committed:
Start Espresso sequencer when running e2e tests in Espresso mode

jbearer committed:
Fix connection between test sequencer nodes and test Geth

jbearer committed:
Implement espresso.Client

With this change, I can run TestSystemE2E and see Espresso batches
successfully get sequenced, although the test later crashes in
the derivation pipeline on the unimplemented `CheckBatchEspresso`.
Also, there are no transactions in the batches, because I need to
change the test to use the proxy RPC as its L2 client, so that
transaction submissions go to Espresso.

Closes #30

jbearer commented:
Adding the `espresso.Client` implementation to this PR since I need it to run the tests and I need the tests to check the client

jbearer committed:
Replace panic in CheckBatchesEspresso stub with acceptance, so I can test

jbearer committed:
Add op-geth-proxy service to e2e Espresso tests

jbearer committed:
Fixes for e2e test setup

jbearer committed:
Improve sequencer logging

jbearer committed:
Simplified unmarshalling for NMT proofs

jbearer committed:
Update Espresso block header

jbearer commented:
Current status: I'm running an end-to-end test in Espresso mode with

```bash
cd op-e2e
OP_E2E_USE_HTTP=true go test -v \
                -l1-allocs ../.devnet-espresso/allocs-l1.json \
                -l1-deployments ../.devnet-espresso/addresses.json \
                -deploy-config ../packages/contracts-bedrock/deploy-config/devnetL1-espresso.json \
                -run TestSystemE2E ./... 2>&1 | tee /tmp/log.txt
```

The first thing this test does is force a deposit through L1, without sending it to the sequencer. This actually works, which is awesome! It means the L1-L2 bridge/escape hatch is going to work after this integration.

Next it sequences an L2 transaction by sending it to the sequencer RPC (which I have configured to point at the op-geth-proxy service). From logs, I can see that the transaction is sequenced by HotShot and included in a batch by the OP sequencer, which is then sent to op-geth and sealed. Also awesome!

However, the test gets stuck immediately after that, because the next HotShot block includes the same transaction, whose nonce is of course invalid the second time. op-geth refuses to execute a block with an invalid transaction, and the sequencer gets stuck in a loop of trying to seal the same batch repeatedly until the test times out. So, it is looking like we'll need to do something about invalid transactions :(

jbearer commented:
Found the place where they add an exception for deposit transactions, where invalid transactions are converted to failed transactions and still added to the block: https://github.com/ethereum-optimism/op-geth/blob/optimism/core/state_transition.go#L355.

We'll have to fork op-geth, but I think adding a similar exception for Espresso transactions should fix the problem. It makes sense that these would be treated similarly, as they both have the same problem: they are force-included (deposit by L1, Espresso transactions by HotShot), the sequencer has no discretion to discard invalid ones.

nomaxg commented:
> Found the place where they add an exception for deposit transactions, where invalid transactions are converted to failed transactions and still added to the block: https://github.com/ethereum-optimism/op-geth/blob/optimism/core/state_transition.go#L355.
> 
> We'll have to fork op-geth, but I think adding a similar exception for Espresso transactions should fix the problem. It makes sense that these would be treated similarly, as they both have the same problem: they are force-included (deposit by L1, Espresso transactions by HotShot), the sequencer has no discretion to discard invalid ones.

We could also have the sequencer remove failed transactions from the payload by repeatedly shrinking the txns payload until the engine succeeds and somehow submit failed txn indices to the batch inbox so that validators can deterministically remove the same transactions before executing. However, this adds a lot of overhead to the pipeline so forking op-geth may be the better approach.

jbearer commented:
Yeah, I think of this as a last resort. The derivation pipeline would also have to do the same thing, execute the full batch and gradually shrink it, to determine that the sequencer shrank it honestly.

I don't think it will be _too_ much work to at least see if patching op-geth is feasible. I'll try to find out today.

jbearer committed:
Point at fork of op-geth

jbearer committed:
Point at forked Geth

Our fork of op-geth ignores invalid transactions, and as a result
TestSystemE2E now passes locally (it won't pass in CI until
https://github.com/EspressoSystems/espresso-sequencer/pull/570 is
merged and the new Docker images are built).

Next step is to fix up the existing CI jobs if I broke them, and
figure out which other e2e tests I can enable in Espresso mode.

jbearer committed:
Ensure Docker images are build before running e2e tests

jbearer committed:
Merge remote-tracking branch 'origin/integration' into test/e2e-espresso

jbearer committed:
Attempt to fix Circle CI config

jbearer committed:
Attempt 2

jbearer committed:
Fix Makefile

jbearer committed:
Install forge for e2e tests

jbearer committed:
Attempt 3

jbearer committed:
Handle rejected transactions

jbearer committed:
Make Espresso e2e setup more reliable

We start up the Espresso system and wait for it to show progress
before starting the L2 chains. Otherwise, the L2 chain is initialized
with the L1 origin from before we start up Espresso, and by the
time Espresso is up 10ish seconds later, the L2's view of L1 is
way behind Espresso, which causes tests to time out while the L2
is catching up.

The new setup process more closely matches the deployment order for
a real system:
- The L1 exists first
- Espresso is deployed
- OP chains are deployed in Espresso mode (or opt into Espresso mode)
  _after_ the Espresso ledger already exists

jbearer committed:
Fix lints

jbearer committed:
Build before running tests

jbearer committed:
Run e2e tests with original Docker image, but install docker-compose

jbearer committed:
Install docker-compose without sudo

jbearer committed:
For debugging, add a manual trigger for e2e tests only

jbearer committed:
Add debugging output for e2e jobs

jbearer committed:
It seems CI jobs run as root, so use recommended docker-compose installation just without sudo

jbearer committed:
Committed the wrong file

jbearer committed:
Install Docker CLI for e2e tests

jbearer committed:
Install Docker using apt-get

jbearer committed:
We are root, so don't use sudo

jbearer committed:
Non-interactive apt-get install

jbearer committed:
I guess apt-get command didn't install docker-compose

jbearer committed:
Fix nil check

jbearer committed:
Increase timeouts in TestERC20BridgeDeposits so we can run it with Espresso

jbearer committed:
Run selected e2e Espresso tests

jbearer committed:
Configure engine to use Espresso mode when appropriate

jbearer commented:
A bunch of tests are now passing. Many are not, but I think at this point the best thing to do is clean up CI and then merge it, so we can move on to working on the demo and cleaning up some loose ends. We can selectively fix and enable more tests time permitting. I suspect many of the failing tests are just having timeouts, since there's a little bit of extra latency in Espresso mode compared to the centralized sequencer.

jbearer committed:
Add quotes for bash command

jbearer committed:
Don't run Espresso e2e tests in parallel, the runner can't handle it

jbearer committed:
Better error messages for startup failures

jbearer committed:
Move e2e job to bare metal to avoid issues with nested Docker

jbearer committed:
Handle race condition in TestSystemRPCAltSync

jbearer committed:
Install gotestsum in VM executor

jbearer committed:
Enable additional Espresso end-to-end test

jbearer committed:
Wrap every error return in e2e startup, so I can see where it's failing in CI

jbearer committed:
Decouple Espresso e2e tests from legacy ones, so I can just test Espresso

jbearer committed:
More logging when waiting for block height

jbearer committed:
Dump Espresso logs on startup failure

jbearer committed:
Bind services to 0.0.0.0 instead of localhost, so they can be reached from containers

jbearer committed:
Enable verbose logging in sequencer for e2e tests

jbearer committed:
Fix logging

jbearer committed:
Yet more logging

jbearer committed:
Increase timeouts for Espresso mode e2e tests

jbearer committed:
Increase more timeouts

jbearer committed:
Ensure JSON parsing fails if required fields are missing

jbearer committed:
Add mock NMT validation interface

jbearer committed:
Ready for review

* Put all CI jobs back in the same dependency graph
* Turn sequencer logging back to INFO level

jbearer committed:
Fix attributes queue test

jbearer committed:
Update op-geth

jbearer committed:
Merge remote-tracking branch 'origin/integration' into test/e2e-espresso

jbearer committed:
Fix typo

jbearer committed:
Fix bug in BlockToBatch

nomaxg commented on go.mod:
```diff
@@ -198,6 +199,6 @@ require (
+replace github.com/ethereum/go-ethereum => github.com/EspressoSystems/op-geth v1.2.0-espresso.0.20230829005208-4111f649c559
 
-//replace github.com/ethereum/go-ethereum v1.12.0 => ../go-ethereum
+//replace github.com/ethereum/go-ethereum v1.12.0 => ../op-geth
```
remove this comment?

nomaxg commented on op-e2e/setup.go:
```diff
@@ -236,12 +246,148 @@ func (sys *System) Close() {
+	}
+}
+
+func (e *EspressoSystem) StartGethProxy(sequencer *node.Node) error {
```
Nice, we should add a real hotshot commitment task/contract deployment to this system eventually

nomaxg commented on op-e2e/setup.go:
```diff
@@ -236,12 +246,148 @@ func (sys *System) Close() {
+	}
+}
+
+func dockerComposePort(projectName string, composeFile string, service string, internalPort uint16) (uint16, error) {
```
nit: DockerComposePort

jbearer replied:
I think lower case is correct because it's private to this file? At least that is what I have gathered is the convention

nomaxg reviewed (approved):
LGTM

jbearer committed:
Remove commented out line from go.mod

jbearer committed:
Point at tagged op-geth

jbearer commented:
I'm pretty sure the failing test is just flaky. It passes sometimes, and when it fails it's just a timeout. And it passes consistently for me locally. Since it's a legacy mode test I don't think it's worth debugging right now, I'm just going to merge this
